tests/services/ci_manager_test.py

```diff
@@ -0,0 +1,357 @@
+            repo_owner=job.repo_owner,
+            repo_name=job.repo_name,
+            ref=job.ref,
+        )
+        for job in queued_jobs
+    ]
+
+    # Wait for the first two jobs to start processing
+    await asyncio.gather(
+        *[lifecycle.processing.wait() for lifecycle in completed_lifecycles]
+    )
+
+    # Wait for the first two jobs to complete
+    for job in completed_jobs:
+        job.proceed_event.set()
+    await asyncio.gather(
+        *[lifecycle.processed.wait() for lifecycle in completed_lifecycles]
+    )
+
+    # Wait for the next two jobs to start processing
+    await asyncio.gather(
+        *[lifecycle.processing.wait() for lifecycle in in_progress_lifecycles]
+    )
+
+    # We'll check this later
+    summary = ci_manager.summarize()
+
+    # Tell the CI manager to stop and let the next two jobs proceed only when
+    # we know the shutdown process has reached a certain point
+    task = asyncio.create_task(ci_manager.aclose())
+    await ci_manager.lifecycle.marked_remaining.wait()
+
+    for job in in_progress_jobs:
+        job.proceed_event.set()
+    await task
+
+    # Make sure the jobs that began processing finished processing
+    assert all(
+        lifecycle.processed.is_set() for lifecycle in in_progress_lifecycles
+    )
+
+    # Make sure the queued jobs never procsesed
+    assert not any(
+        lifecycle.processed.is_set() for lifecycle in queued_lifecycles
+    )
+    assert not any(
+        lifecycle.processing.is_set() for lifecycle in queued_lifecycles
+    )
+
+    # Check the point-in-time summary. It's possible for either worker to
+    # have either in_progress job.
+    expected_summary1 = CiManagerSummary(
+        workers=[
+            CiWorkerSummary(
+                user=User(username="user1", uidnumber=None, gidnumber=None),
+                num_processed=1,
+                current_job=CiJobSummary(
+                    commit_url=HttpUrl(
+                        "https://github.com/repo_owner_in_progress_1/repo_name_in_progress_1/commit/ref_in_progress_1"
+                    )
+                ),
+            ),
+            CiWorkerSummary(
+                user=User(username="user2", uidnumber=None, gidnumber=None),
+                num_processed=1,
+                current_job=CiJobSummary(
+                    commit_url=HttpUrl(
+                        "https://github.com/repo_owner_in_progress_2/repo_name_in_progress_2/commit/ref_in_progress_2"
+                    )
+                ),
+            ),
+        ],
+        num_queued=2,
+    )
+
+    expected_summary2 = CiManagerSummary(
+        workers=[
+            CiWorkerSummary(
+                user=User(username="user1", uidnumber=None, gidnumber=None),
+                num_processed=1,
+                current_job=CiJobSummary(
+                    commit_url=HttpUrl(
+                        "https://github.com/repo_owner_in_progress_2/repo_name_in_progress_2/commit/ref_in_progress_2"
```

Comment:
Nit: You can break long strings over multiple lines and keep the indentation by putting multiple quoted strings on successive lines without any commas or the like between them and rely on Python implicit string concatenation. That makes this sort of thing a bit easier to read.